include/databaseFunctions_blog_pic.php

```diff
@@ -1,58 +1,450 @@
+		echo"<form method='post'>";
+		ShowTextField(true, 'username');
+		ShowPasswordField('password', 'password');
+		echo"<input type='submit' name = 'login' value='Login'>
+			</form>";
+	}
+
+	function IsLoggedIn(){
+		echo"";
+		if(isset($_SESSION['userID'])){
+			return true;
+		}
+		return false;
+	}
+
+	function PersonalHeading(){
+		echo"
+			<p class='personal'>hi ".$_SESSION['username']."</p>
+		";
+		Logout();
+	}
+
+	function Logout(){
+		if(isset($_REQUEST['logout'])){
+			// echo"<script>
+			// 		alert('Are you sure you want to log out?')
+			// 	</script>
+			// ";
+			session_destroy();
+			header("Location: /index.php");
+			exit();
+		}
+		echo"
+			<form>
+				<input class='personal' type='submit' name='logout' value='Log Out'>
+			</form>
+		";
+	}
+
+	function ValidDelete($postID){
+		// echo"validdelete called";
+		// var_dump($_SESSION);
+		if(isset($_SESSION['userID'])){
+			if($_SESSION['userID'] == GetPostCreator($postID)){
+				return true;
+			}
+			if(GetUser($_SESSION['username'])['userType'] == 'admin'){
+				return true;
+			}
+		}
+		if(GetPostCreator($postID) == NULL){
+			return true;
+		}
+		return false;
+	}
+
+	//TAG DATABASE FUNCTIONS
+	function DeleteTagFromPost($tagname, $postID){		//CHECK DBQUERY SYNTAX
+		$result = dbQuery("
+			DELETE FROM posttags
+			WHERE (SELECT tagname
+				FROM tags
+				INNER JOIN posttags ON
+				tags.tagID = posttags.tagID) = :tagname
+			AND postID = :postID
+		", array('tagname'=>$tagname, 'postID'=>$postID))->fetch();
+	}
+
+	function DeleteTagOverall($name){
+		$result = dbQuery("
+			DELETE FROM tags
+			WHERE tagname = :name
+		", array('name'=>$name))->fetch();
+	}
+
+	function NewTag($name){
+		dbQuery("
+			INSERT INTO tags (tagname)
+			VALUES(:name)
+		", array('name'=>$name))->fetch();
+	}
+
+	function AttachTags($postID, $tagarray){
+		// echo"attachtags called, postID: ";
+		// var_dump($postID);
+		for($i=0;$i<sizeof($tagarray);$i++){
+			// die("forloop");
+			if(($tagarray[$i] == 'null')||($tagarray[$i] == '')){
+				continue;
+			}
+			if(!TagExists($tagarray[$i])){
+				echo"<br>CREATING A NEW TAG<br>";
+				NewTag($tagarray[$i]);
+			}
+			$tagID=GetTag($tagarray[$i])['tagID'];
+			dbQuery("
+				INSERT INTO posttags (postID, tagID)
+				VALUES(:postID, :tagID)
+			", array('postID'=>$postID, 'tagID'=>$tagID))->fetchAll();
+		}
+		 // die();
+	}
+
+	function GetTag($name){
+		$result=dbQuery("
+			SELECT *
+			FROM tags
+			WHERE tagname = :name
+		", array('name'=>$name))->fetch();
+		return $result;
+	}
+
+	function GetAllTags($postID){		//CONFIRM SYNTAX OF DBQUERY
+		$result=dbQuery("
+			SELECT tagname
+			FROM tags
+			INNER JOIN posttags ON tags.tagID = posttags.tagID
+			WHERE posttags.postID = :postID
+		", array('postID'=>$postID))->fetchAll();
+		// var_dump($result);
+		return $result;
+	}
+
+	function HasTags($postID){
+		// echo"HasTags called";
+		$result = dbQuery("
+			SELECT *
+			FROM posttags
+			WHERE EXISTS
+			(SELECT 1 FROM posttags
```

Comment:
This isn't the standard way to write this. You can just say `WHERE postID = :postID` and skip the subquery. It's pretty rare that you need to use a subquery in a SELECT.

Having said that, it might be easier to just have a function called `GetTagsForPost()` and return all the tags linked to this post, and then you don't need this function.